rust/arrow/src/compute/kernels/sort.rs

```diff
@@ -0,0 +1,295 @@
+            sort_primitive::<Time32MillisecondType>(values, v, n, &options)
+        }
+        DataType::Time64(Microsecond) => {
+            sort_primitive::<Time64MicrosecondType>(values, v, n, &options)
+        }
+        DataType::Time64(Nanosecond) => {
+            sort_primitive::<Time64NanosecondType>(values, v, n, &options)
+        }
+        DataType::Timestamp(Second, _) => {
+            sort_primitive::<TimestampSecondType>(values, v, n, &options)
+        }
+        DataType::Timestamp(Millisecond, _) => {
+            sort_primitive::<TimestampMillisecondType>(values, v, n, &options)
+        }
+        DataType::Timestamp(Microsecond, _) => {
+            sort_primitive::<TimestampMicrosecondType>(values, v, n, &options)
+        }
+        DataType::Timestamp(Nanosecond, _) => {
+            sort_primitive::<TimestampNanosecondType>(values, v, n, &options)
+        }
+        DataType::Interval(IntervalUnit::YearMonth) => {
+            sort_primitive::<IntervalYearMonthType>(values, v, n, &options)
+        }
+        DataType::Interval(IntervalUnit::DayTime) => {
+            sort_primitive::<IntervalDayTimeType>(values, v, n, &options)
+        }
+        DataType::Duration(TimeUnit::Second) => {
+            sort_primitive::<DurationSecondType>(values, v, n, &options)
+        }
+        DataType::Duration(TimeUnit::Millisecond) => {
+            sort_primitive::<DurationMillisecondType>(values, v, n, &options)
+        }
+        DataType::Duration(TimeUnit::Microsecond) => {
+            sort_primitive::<DurationMicrosecondType>(values, v, n, &options)
+        }
+        DataType::Duration(TimeUnit::Nanosecond) => {
+            sort_primitive::<DurationNanosecondType>(values, v, n, &options)
+        }
+        DataType::Utf8 => sort_string(values, v, n, &options),
+        t @ _ => Err(ArrowError::ComputeError(format!(
+            "Sort not supported for data type {:?}",
+            t
+        ))),
+    }
+}
+
+/// Options that define how sort kernels should behave
+#[derive(Clone)]
+pub struct SortOptions {
+    /// Whether to sort in descending order
+    descending: bool,
+    /// Whether to sort nulls first
+    nulls_first: bool,
+}
+
+impl Default for SortOptions {
+    fn default() -> Self {
+        Self {
+            descending: false,
+            nulls_first: false,
+        }
+    }
+}
+
+/// Sort primitive values
+fn sort_primitive<T>(
+    values: &ArrayRef,
+    value_indices: Vec<usize>,
+    null_indices: Vec<u32>,
+    options: &SortOptions,
+) -> Result<UInt32Array>
+where
+    T: ArrowPrimitiveType,
+    T::Native: std::cmp::Ord,
+{
+    let values = values.as_any().downcast_ref::<PrimitiveArray<T>>().unwrap();
+    // create tuples that are used for sorting
+    let mut valids = value_indices
+        .into_iter()
+        .map(|index| (index as u32, values.value(index)))
+        .collect::<Vec<(u32, T::Native)>>();
+    let mut nulls = null_indices;
+    if !options.descending {
```

Comment:
I think this would be more readable if the if were inverted to remove the `!` from the condition?

```rust
if options.decending {
} else {
}
```

Reply:
In fact, I think I would prefer the options to have a field named `ascending` rather than `descending` since ascending is typically the default.

Reply:
My justification for using `descending` instead of `ascending` is that the default behaviour should not require passing options. As boolean's default is false, I'd rather the caller opt in explicitly to sorting in a descending order.
I could set the default to ascending, but I didn't like the below as it'd be less obvious:

```rust
impl Default for SortOptions {
    fn default() -> Self {
        Self {
            ascending: true,
            nulls_first: false,
        }
    }
}
```